ouroboros-consensus/src/unstable-consensus-testlib/Test/Util/QuickCheck.hs

```diff
@@ -72,6 +73,32 @@ checkShrinker p =
 checkInvariant :: (a -> Except String ()) -> (a -> Property)
 checkInvariant f = expectRight () . runExcept . f
 
+-- | Explicit quantification using the “gen-run-shrink-check” pattern.
+--
+-- Instead of the usual two stages where one generates an input and then checks
+-- the property for that input, we rely on three stages: one generates an input,
+-- then transforms it into an output, and then checks the output.
+--
+-- When adding a shrinker to the mix, we can allow it to inspect the output
+-- value as well, which increases its expressivity. This makes sense if the
+-- “run” phase is particularly expensive.
+forAllGenRunShrinkCheck ::
+  Testable prop =>
+  Gen input ->
+  (input -> output) ->
+  (input -> output -> [input]) ->
+  (input -> output -> prop) ->
+  Property
+forAllGenRunShrinkCheck gen run shrink_ check =
+  forAllBlind gen $ \input ->
+    shrinking
+      (run' <.> uncurry shrink_)
```

Comment:
I still find `map run' . uncurry shrink_` or `(\(i, o) -> map run' (shrink i o))` easier to grok.

Reply:
Sorry, I extracted the commits from the other PR but forgot to also apply this comment. No strong opinion on what's better; I took your first suggestion in 27f63dec1ee3f4fb14e7682dea46b65cd365b2f5.